index.rst

```diff
@@ -1,61 +1,420 @@
+- Whether to allow cookie authentication
+- Whether to prohibit ``POST`` with form submission ``Content-Type`` values
+- What ``Content-Security-Policy`` header to add, if any
+
+This can all be managed with manually-written NGINX ingress annotations with each service, with many of the parameters embedded in the ``auth-url`` URL, but this is tedious and error-prone.
+
+We therefore plan to implement a custom Kubernetes resource that specifies a Gafaelfawr-protected resource, and a custom resource controller that writes the ``Ingress`` resource based on that custom resource.
+This will allow the above information to be expressed in more human-readable YAML or derived from the cluster Gafaelfawr configuration.
+
+Isolation
+=========
+
+Web security is closely tied to origins, which are (roughly) defined as the tuple of protocol, hostname, and port.
+All JavaScript running on the same origin (including any URL path in that origin) has effectively the same privileges, can freely make HTTP requests of any type to other URLs on the same origin, can access any local state tied to that origin, and so forth.
+
+Cookies have a similar concept called a site, which controls when the browser will send the cookie alongside a request.
+The site of a cookie is somewhat less restricted than a JavaScript origin and can include subdomains depending on the cookie configuration.
+
+The effect of this security model is that it is not possible to secure web applications against malicious JavaScript running in the same origin.
+To isolate one web application from another, they must run in separate origins.
+
+The easiest way to do this is to give every application a separate origin, usually by changing the hostname.
+In the case of the Science Platform, this would mean assigning a separate hostname to every application, and then using either multiple TLS certificates, a wildcard TLS certificate, or a TLS certificate with multiple :abbr:`SANs (Subject Alternative Names)`.
+
+However, separate origins are only crucial for web applications that run JavaScript in the browser.
+REST API endpoints can safely share the same origin provided that they do not need to support cross-origin requests, do not serve pages or JavaScript that run in a web browser, and do not share an origin with a web application that does so.
+Having all Science Platform REST APIs share the same origin is useful for documentation purposes and allows more flexibility about which API endpoints are served by which backend implementations.
+
+Therefore, the isolation plan for the Rubin Science Platform is:
+
+- Serve the Notebook Aspect spawning interface from its own origin
+- Serve each user's notebook from a per-user origin (see :ref:`jupyterlab-origin`)
+- Serve the Portal Aspect from its own origin
+- Serve the authentication system from its own origin
+- Serve all APIs from a single origin shared by the APIs, but separate from the other origins
```

Comment:
Good catch, I wasn't thinking of that. I think this is worth explicitly calling out since there's also a `Content-Security-Policy` header that should be added for such APIs. I've added some additional wording for that.